hash.go

```diff
@@ -0,0 +1,112 @@
+package bn256
+
+// Hash implements a hashing function into the G1 group. It uses Fouque-Tibouchi encoding described here
+// https://tools.ietf.org/pdf/draft-irtf-cfrg-hash-to-curve-04.pdf
+func Hash(msg []byte) *G1 {
+	return mapToCurve(hashToBase(msg))
+}
+
+func mapToCurve(t *gfP) *G1 {
+	// calculate w = (s * t)/(1 + B + t^2)
+	w := &gfP{}
+
+	// s is a square root of -3
+	s := &gfP{0x236e675956be783b, 0x053957e6f379ab64, 0xe60789a768f4a5c4, 0x04f8979dd8bad754}
+
+	t2 := &gfP{}
+	gfpMul(t2, t, t)
+	gfpAdd(w, curveB, t2)
+	gfpAdd(w, w, &one)
+	w.Invert(w)
+
+	gfpMul(w, w, s)
+	gfpMul(w, w, t)
+
+	e := legendre(t)
+	cp := &curvePoint{z: one, t: one}
+
+	// calculate x1 = ((-1 + s) / 2) - t * w
```

Comment:
Note that for calculating `x1` the first part of the sum is a constant which can be precomputed. Thus, you do not need to compute the division by two.

Reply:
done.

Reply:
done.